src/http_client_async.zig

```diff
@@ -2054,14 +2124,52 @@ pub fn onData(this: *HTTPClient, comptime is_ssl: bool, incoming_data: []const u
     }
 }
 
-fn fail(this: *HTTPClient, err: anyerror) void {
+pub fn closeAndAbort(this: *HTTPClient, reason: JSC.JSValue, comptime is_ssl: bool, socket: NewHTTPContext(is_ssl).HTTPSocket) void {
+    socket.ext(**anyopaque).?.* = bun.cast(
+        **anyopaque,
+        NewHTTPContext(is_ssl).ActiveSocket.init(&dead_socket).ptr(),
+    );
+    if (this.globalThis) |globalThis| {
+        const reason_str = reason.toString(globalThis).getZigString(globalThis);
+
+        if (reason_str.len >= 12) {
+            var type_str = reason_str.substr_slice(0, 10);
+            const ABORT_ERROR = JSC.ZigString.init("AbortError");
+            if (type_str.eql(ABORT_ERROR)) {
+                const message_str = reason_str.substring(12);
+                const exception = JSC.AbortSignal.createAbortError(&message_str, &JSC.ZigString.Empty, globalThis);
+                this.fail(error.Canceled, exception);
+            } else {
+                type_str = reason_str.substr_slice(0, 12);
+                const TIMEOUT_ERROR = JSC.ZigString.init("TimeoutError");
+                if (type_str.eql(TIMEOUT_ERROR)) {
+                    const message_str = reason_str.substring(14);
+                    const exception = JSC.AbortSignal.createTimeoutError(&message_str, &JSC.ZigString.Empty, globalThis);
+                    this.fail(error.Canceled, exception);
+                } else {
+                    const exception = JSC.AbortSignal.createAbortError(&reason_str, &JSC.ZigString.Empty, globalThis);
+                    this.fail(error.Canceled, exception);
+                }
+            }
+        } else {
+            const exception = JSC.AbortSignal.createAbortError(&reason_str, &JSC.ZigString.Empty, globalThis);
+            this.fail(error.Canceled, exception);
+        }
+    } else {
+        this.fail(error.Canceled, null);
+    }
+
+    socket.close(0, null);
```

Comment:
also we should close the socket first imo

that way if anything happens as a result of the rest of this code we always make sure to close the socket first